File: filler_arena/filler_arena/board.py
from copy import deepcopy
from random import randint


class Board:
    def __init__(self, size):
        self.size = size
        self.board = [[None for _ in range(0, self.size)] for _ in range(0, self.size)]

    def __is_valid_box(self, y, x, pval):
        return (x >= 0 and x < self.size and y >= 0 and y < self.size
            and (not self.board[y][x] or self.board[y][x] == pval))

    def insert_piece(self, piece, y, x, player):
        neighbor_cnt = 0
        board = deepcopy(self.board)
        board = [[box * -1 if box and box < 0 else box for box in row] for row in board]
        for py in range(0, piece.size):
            for px in range(0, piece.size):
                if not piece.piece[py][px]:
                    continue
                if not self.__is_valid_box(py + y, px + x, player.value):
                    return False
                if board[y + py][x + px] == player.value:
                    neighbor_cnt += 1
                board[y + py][x + px] = -player.value
        if neighbor_cnt == 1:
            self.board = board
            return True
        return False

    def set_start_position(self, y=None, x=None):
        if not y or not x:
            y = randint(0, self.size - 1)
            x = randint(0, self.size - 1)
        self.board[y][x] = 1
        y = -y if y else self.size - 1
        x = -x if x else self.size - 1
        self.board[y][x] = 2

File: v4/arena/src/game/board.rs
use serde::{Serialize, Deserialize};

use super::piece::Piece;

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Board {
    pub size: usize,
    board: Vec<Vec<i8>>,
}


impl Board {
    pub fn new(size: usize) -> Self {
        let board = vec![vec![0; size]; size];
        Board { size, board }
    }

    fn is_valid_box(&self, y: isize, x: isize, player: i8) -> bool {
        y >= 0
            && x >= 0
            && y < self.size as isize
            && x < self.size as isize
            && (self.board[y as usize][x as usize] == 0
                || self.board[y as usize][x as usize] == player)
    }


    pub fn is_valid_piece_coord(&self, piece: &Piece, y: isize, x: isize, player: i8) -> bool {
        let mut neighbor_cnt = 0;

        for py in 0..piece.size {
            for px in 0..piece.size {
                if piece.piece[py][px].is_none() {
                    continue;
                }

                if !self.is_valid_box(y + py as isize, x + px as isize, player) {
                    return false;
                }

                if self.board[(y + py as isize) as usize][(x + px as isize) as usize] == player {
                    neighbor_cnt += 1;
                }
            }
        }
        if neighbor_cnt == 1 {
            return true;
        }
        return false;
    }

    pub fn set_piece(&mut self, piece: &Piece, y: isize, x: isize, player: i8) -> bool {
        for py in 0..piece.size {
            for px in 0..piece.size {
                if piece.piece[py][px].is_none() {
                    continue;
                }
                if self.board[(y + (py as isize)) as usize][(x + (px as isize)) as usize] <= 0 {
                    self.board[(y + (py as isize)) as usize][(x + (px as isize)) as usize] = player;
                }
            }
        }
        return true;
    }

    pub fn set_start_position(&mut self, y: usize, x: usize) {
        self.board[y][x] = 1;
        let y = if y > 0 { self.size - y } else { self.size - 1 };
        let x = if x > 0 { self.size - x } else { self.size - 1 };
        self.board[y][x] = 2;
    }

    pub fn display(&self) {
        for row in &self.board {
            for cell in row {
                print!("{}", cell);
            }
            println!();
        }
    }
}
